File: 1/2021.03.28/341.Flatten Nested List Iterator/java/solution.java
public class NestedIterator implements Iterator<Integer> {

    LinkedList<NestedInteger> list;

    public NestedIterator(List<NestedInteger> nestedList) {
        list = new LinkedList<>(nestedList);
    }

    @Override
    public Integer next() {
        return list.remove(0).getInteger();
    }

    @Override
    public boolean hasNext() {
        while (!list.isEmpty() && !list.get(0).isInteger()) {
            List<NestedInteger> remove = list.remove(0).getList();
            for (int i = remove.size() - 1; i >= 0; i--) {
                list.addFirst(remove.get(i));
            }
        }
        return !list.isEmpty();
    }
}

File: 1/2021.03.28/341.Flatten Nested List Iterator/python/solution.py
class NestedIterator:
    def __init__(self, nestedList: [NestedInteger]):
        self.res = []
        for child in nestedList:
            self.DFS(child)
        self.size = len(self.res)
        self.index = -1
    
    def next(self) -> int:
        self.index += 1
        return self.res[self.index]
    
    def hasNext(self) -> bool:
        return self.index + 1 < self.size

    def DFS(self, root):
        if root.isInteger():
            self.res.append(root.getInteger())
            return
        for child in root.getList():
            self.DFS(child)
